Add registry doc (#512)
awslabs/soci-snapshotter: docs/registry.md

 # Registry Support for SOCI

> SOCI is supported by most registries. The notable exception to this is Duckerhub, which currently does not support subject fields in image manifests.

Reviewer: Can we rephrase all the various uses of "registry supports soci" to refer more to the features we depend on and less to us specifically? This phrasing suggests that we want registries to do something special to support us in particular. I want to emphasize that we rely on interoperable standards. Perhaps something like "The features that SOCI relies on are available in most registries" here, but less verbose rephrasing in other instances further down the doc. Perhaps "is compatible with" rather than "supports" or "is supported by".

Author: That's a good call out. I'm going to change "support" to "compatible with" when talking about SOCI and only use the term "support" in reference the specific OCI spec features that we rely on.

Author: Can you take another look at let me know if you like the new phrasing?

Reviewer: Looks great. Love the phrasing of it as registries supporting features rather than us supporting registries.

> SOCI is supported by most registries. The notable exception to this is Duckerhub, which currently does not support subject fields in image manifests.

Reviewer 2: is duckerhub a thing? or is this a dockerhub typo?

Reviewer: ```suggestion
SOCI is supported by most registries. The notable exception to this is Docker Hub, which currently does not support subject fields in image manifests.
```

### Fallback

If the referrers API is not available, the OCI distribution spec defines a [fallback mechanism that works with existing registries](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api). In this mechanism, the contents that would normally be returned by the referrers API are instead put into an [OCI Image Index](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/image-index.md) which is tagged in the registry with the digest of the manifest to which it refers.

For example, imagine you had an image `myregistry.com/image:latest` with digest `sha:123`. If you created and pushed a SOCI index for that image, there would also be a new image index `myregistry.com/image:sha-123` which contains the SOCI index' descriptor. At runtime, the soci-snapshotter will pull the `myregistry.com/image:sha-123` index and apply client side filtering to discover the SOCI index.

> For clarity in the rest of this section, we will refer to `myregistry.com/image:sha-123` as the "fallback" (as opposed to image index) to distinguish it from the SOCI index. Please note that this is not standard terminology.

Reviewer: I suggest finding a way to make/use standard terminology here.

Author: "referrers tag schema" is the closest thing to standard terminology, which I think is more confusing that just using the term "fallback".

An important note here is that the fallback is managed on the *client side* by the tool performing the push. There is therefore a race condition when pushing a SOCI index because the fallback has to be pulled, modified to add the new SOCI index, and then pushed back to the registry. If a second artifact is pushed that references the same image digest, then one modification of the fallback could clobber the other.

> To clarify the scope of this problem, the fallback is unique per image digest. Multiple artifacts (SOCI Indices, signatures, etc.) can modify the same fallback. The image digest is generally unique per image/platform pair. As an example of what this means in practice, concurrently creating a SOCI index for the an image for platforms `linux/amd64` and `linux/i386` is safe because the image digests will be different. Concurrently creating a SOCI index and signature for an image and platform `linux/amd64` is unsafe because both artifacts will refer to the same image digest.

Reviewer: `for the an image`

> To clarify the scope of this problem, the fallback is unique per image digest. Multiple artifacts (SOCI Indices, signatures, etc.) can modify the same fallback. The image digest is generally unique per image/platform pair. As an example of what this means in practice, concurrently creating a SOCI index for the an image for platforms `linux/amd64` and `linux/i386` is safe because the image digests will be different. Concurrently creating a SOCI index and signature for an image and platform `linux/amd64` is unsafe because both artifacts will refer to the same image digest.

Reviewer: ```suggestion
To clarify the scope of this problem, the fallback is unique per image digest. Multiple artifacts (SOCI Indices, signatures, etc.) can modify the same fallback. The image digest is generally unique per image/platform pair. As an example of what this means in practice, concurrently creating a SOCI index for the image for platforms `linux/amd64` and `linux/i386` is safe because the image digests will be different. Concurrently creating a SOCI index and signature for an image and platform `linux/amd64` is unsafe because both artifacts will refer to the same image digest.
```

| latest      | Image       | The actual image                                                                                           |
| \<untagged> | Image       | The SOCI index manifest. This may appear as type SOCI Index or Other                                       |
| sha:123     | Image Index | The fallback image index. This will only be present for registries which do not support the referrers API. |

## List of Registry Support

Registries that are not listed have not been tested by the SOCI maintainers or reported by the community, but they may still support SOCI.

| Registry                                                                                  | Supported? | Mechanism     | Notes                                                |
| ----------------------------------------------------------------------------------------- | ---------- | ------------- | ---------------------------------------------------- |
| [Docker Hub](https://hub.docker.com)                                                      | No         | N/A           | Does not support image manifests with subject fields |

> | [Amazon Elastic Container Registry (ECR)](https://aws.amazon.com/ecr/)                    | Yes        | Fallback      |                                                      |

Reviewer 2: Does ECR public support?

| [Google Cloud Artifact Registry (AR)](https://cloud.google.com/artifact-registry)         | No         | N/A           | Testing the referrers API redirects to login         |
| [Quay.io](https://quay.io)                                                                | No         | N/A           | Does not support image manifests with subject fields |
| [Artifactory](https://jfrog.com/artifactory/)                                             | Yes        | Fallback      |                                                      |
| [Harbor](https://github.com/goharbor/harbor)                                              | Yes        | Fallback      |                                                      |
| [Distribution](https://github.com/distribution/distribution)                              | Yes        | Fallback      |                                                      |
| [OCI-playground Distribution](https://github.com/oci-playground/distribution)             | Yes        | Referrers API |                                                      |

### Failure Examples

> Below are some slightly redacted exmples from the services that don't support SOCI.

Reviewer: ```suggestion
Below are some slightly redacted examples from the services that don't support SOCI.
```

 # Registry Compatibility with SOCI

SOCI is compatible with most registries. To check if your registry of choice is compatible, see [List of Registry Compatibility](#list-of-registry-compatibility).

For most use-cases, compatibility is the only concern. However, there is a difference in *how* registries work with SOCI that could cause surprising edge cases. The rest of this document is a technical dive into how SOCI indices are stored and retrieved from registries and the surprises you might encounter.

## Registry Requirements

In order for a registry to be compatible SOCI it must support the following features of the OCI distribution and image specs:

1) Accept [OCI Image Manifests](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) with [subject fields](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md#image-manifest-property-descriptions) and arbitrary config media types.
This allows the registry to store SOCI indices and is supported by most registries.

> 2) (optional) Support the [OCI referrers API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers)

Reviewer 3: missing a `.` after [OCI referrers API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers)

> This adds convenience around retrieving SOCI indices from the registry. If it is not supported, there is a [fallback mechanism](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api) that can be used instead, but it has a few issues noted in the next section.

Reviewer 3: > there is a [fallback mechanism](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api) that can be used instead

It's unclear to me how this can be used. Is it something a registry needs to support? (i.e., do you mean "If it's not supported, the registry needs to support the fallback mechanism").

| [Google Cloud Artifact Registry (AR)](https://cloud.google.com/artifact-registry)         | No          | N/A           | Testing the referrers API redirects to login         |
| [Quay.io](https://quay.io)                                                                | No          | N/A           | Does not support image manifests with subject fields |
| [Artifactory](https://jfrog.com/artifactory/)                                             | Yes         | Fallback      |                                                      |
| [Harbor](https://github.com/goharbor/harbor)                                              | Yes         | Fallback      |                                                      |
| [Distribution](https://github.com/distribution/distribution)                              | Yes         | Fallback      |                                                      |
| [OCI-playground Distribution](https://github.com/oci-playground/distribution)             | Yes         | Referrers API |                                                      |

> 

Reviewer 3: Unnecessary empty lines (1 empty line is sufficient).

### Failure Examples

Below are some slightly redacted examples from the services that don't support the features needed to be compatible SOCI.

**Docker Hub**

> ```

Reviewer 3: nit: an empty line before and after the code block, also language annotation (e.g., ```shell/console/bash/etc)

$ sudo ./out/soci push --user $USERNAME:$PASSWORD docker.io/####/busybox:latest
checking if a soci index already exists in remote repository...
pushing soci index with digest: sha256:d6ebffd218ead37e4862172b4f19491341e72aebc3cc6d9cf1a22297c40fb3c2
pushing artifact with digest: sha256:cce4c7e12e01b32151d69348fcf52e0db7b44f6df6c23c511fa5c52eaf272c28
pushing artifact with digest: sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
skipped artifact with digest: sha256:acaddd9ed544f7baf3373064064a51250b14cfe3ec604d65765a53da5958e5f5
successfully pushed artifact with digest: sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
successfully pushed artifact with digest: sha256:cce4c7e12e01b32151d69348fcf52e0db7b44f6df6c23c511fa5c52eaf272c28
pushing artifact with digest: sha256:d6ebffd218ead37e4862172b4f19491341e72aebc3cc6d9cf1a22297c40fb3c2
soci: error pushing graph to remote: PUT "https://registry-1.docker.io/v2/####/busybox/manifests/sha256:d6ebffd218ead37e4862172b4f19491341e72aebc3cc6d9cf1a22297c40fb3c2": response status code 404: notfound: not found
```

The index manifest can't be pushed at all.

**Amazon ECR Public Gallery**

> ```

Reviewer 3: same here

```

The index manifest can't be pushed at all.

**Google Cloud Artifact Registry**
```
$ sudo ./out/soci push --user $USERNAME:$PASSWORD us-east1-docker.pkg.dev/####/busybox:latest
checking if a soci index already exists in remote repository...
soci: failed to fetch list of referrers: GET "https://accounts.google.com/v3/signin/identifier?dsh=###&continue=https%3A%2F%2Fconsole.cloud.google.com%2Fartifacts%2Ftags%2Fv2%2Fus-east1%2F####%252Fbusybox%252Freferrers%252Fsha256%2Facaddd9ed544f7baf3373064064a51250b14cfe3ec604d65765a53da5958e5f5...&service=cloudconsole&flowName=WebLiteSignIn&flowEntry=ServiceLogin": failed to decode response: invalid character '<' looking for beginning of value
```

This is before pushing the SOCI index. Looking up an existing referrer is redirecting to an auth flow.

Pushing a SOCI index works as long as the SOCI CLI receives the `--existing-index allow` flag to skip the check for existing indices. This is because Artifact Registry appears to always redirect requests to the referrers API to an auth flow, despite other authenticated requests to the registry succeeding as expected. Pulling always falls back to OverlayFS due to the redirect.

> 

Reviewer 3: nit: 1 empty line

 # Registry Support for SOCI

> SOCI is supported by most registries. The notable exception to this is Duckerhub, which currently does not meet our technical requirements. 

Reviewer 3: "meet our technical requirements" sounds like we have some non-standard requirement. Is it possible to change to something like "which doesn't fully support OCI 1.0 used by SOCI".

> To check if your registry of choice is supported, see [#list-of-registry-support](List of Registry Support).

Reviewer 4: switch the order of link with text [List of...] (# link)

For most use-cases, support is the only concern. However, there is a difference in *how* registries support SOCI that could cause surprising edge cases. The rest of this document is a technical dive into how SOCI indices are stored and retrieved from registries and the surprises you might encounter.

## Registry Requirements

In order for a registry to support SOCI it must have the following features:

> 1) Accept [OCI Image Manifests](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md) with [subject fields](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md#image-manifest-property-descriptions) and arbitrary config media types.

Reviewer 3: Is this part of OCI 1.0 (or prior to 1.0)? If so it might be good to refer a 1.0.0 tag instead of 1.1.0-rc2 which looks like a new feature to be added in OCI 1.1.

Author: I'd like to reference the same version of the spec throughout this doc to reduce confusion. In this case, we actually do need the latest doc which defines the subject field.

Reviewer 3: I'm okay with that. then we need to change v1.1.0-rc1 to v1.1.0-rc2 (both are used in the doc now)

Author: They're actually 2 difference specs. Distribution spec v1.1.0-rc1 and Image spec v1.1.0-rc2.

Reviewer 3: oh missed the name part 😅

This allows the registry to store SOCI indices and is supported by most registries.

2) (optional) Support the [OCI referrers API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers)
This adds convenience around retrieving SOCI indices from the registry. If it is not supported, there is a [fallback mechanism](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api) that can be used instead, but it has a few issues noted in the next section. Since the OCI referrers API was defined in the not-yet-released [OCI Distribution Spec v1.1](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md), very few registries support it.

## Referrers API vs Fallback

> Registries can support retrieving SOCI artifacts either though the [OCI referrers API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers) or a [Fallback mechanism](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api). The referrers API is part of the not-yet-released [OCI Distribution Spec v1.1](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md) so support is limited. The Fallback API is supported by all registries, but has notable edge cases.

Reviewer 3: nit: SOCI artifacts -> SOCI artifacts (index, ztoc).

though -> through

Reviewer 3: The 2nd sentence is repeated with the sentence above "2) (optional)..." (copied below). Suggest to remove the one in "2) (optional).."

"Since the OCI referrers API was defined in the not-yet-released [OCI Distribution Spec v1.1](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md), very few registries support it."

Author: I intentionally duplicated some of the wording so that the requirements section is like the summary version and the comparison is the detailed version. I can agree that it's probably not adding enough value, though. I will remove it.

> The SOCI CLI and soci-snapshotter automatically and seamlessly use the referrers API if the registry supports it or the fallback mechanism otherwise.

Reviewer 3: nit: "and seamlessly" is unnecessary

### Referrers API

> The [referrers API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers) is a registry endpoint where an agent can query for all artifacts that reference a given image digest, optionally filtering by artifact type. This means that when an container is launched, the soci-snapshotter can query this API to find SOCI indices that reference the digest of the image-to-be-launched. The registry indexes artifacts for the referrers API when the artifact is pushed.

Reviewer 3: nit: "the image-to-be-launched" -> "the image used by that container"

Reviewer 3: suggest to switch the 2nd and 3rd sentence. (index happens before query).

also "This means that when" -> "When"

### Fallback

> If the referrers API is not available, the OCI spec authors defined a [fallback mechanism that works with existing registries](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api). In this mechanism, the contents that would normally be returned by the referrers API are instead put into an [OCI Image Index](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/image-index.md) which is tagged in the registry with the digest of the manifest to which it refers.

Reviewer 3: nit: "authors defined" -> "defines"

For example, imagine you had an image `myregistry.com/image:latest` with digest `sha:123`. If you created and pushed a SOCI index for that image, there would also be a new image index `myregistry.com/image:sha-123` which contains the SOCI index' descriptor. At runtime, the soci-snapshotter will pull the `myregistry.com/image:sha-123` index and apply client side filtering to discover the SOCI index. 

For clarity in the rest of this section, we will refer to `myregistry.com/image:sha-123` as the "fallback" (as opposed to image index) to distinguish it from the SOCI index. Please note that this is not standard terminology. 

> An important note here is that the fallback is managed *client side* by the tool performing the push. There is therefore a race condition when pushing a SOCI index because the fallback has to be pulled, modified to add the new SOCI index, and then pushed back to the registry. If a second artifact is pushed that references the same image digest, then one modification of the fallback could clobber the other. 

Reviewer 3: *client side* -> on *client side*

To clarify the scope of this problem, the fallback is unique per image digest. Multiple artifacts (SOCI Indices, signatures, etc.) can modify the same fallback. The image digest is generally unique per image/platform pair. As an example of what this means in practice, concurrently creating a SOCI index for the an image for platforms `linux/amd64` and `linux/i386` is safe because the image digests will be different. Concurrently creating a SOCI index and signature for an image and platform `linux/amd64` is unsafe because both artifacts will refer to the same image digest.

Since the fallback is managed client side, the registry does not know about the relationship between SOCI indices and the fallback. Deleting a SOCI index will not delete or modify the fallback. It is up to the user to make the necessary modifications or deletions of the fallback when deleting a SOCI index from the registry.

> ## How SOCI Indices appear to registries

Reviewer 3: make header capitalization consistent
